src/test/tools/obj_verify/obj_verify.c

```diff
@@ -0,0 +1,244 @@
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * obj_verify.c -- tool for creating and verifying a pmemobj pool
+ */
+
+#include <stddef.h>
+#include <fcntl.h>
+
+#include "libpmemobj.h"
+#include "set.h"
+
+#define SIGNATURE_LEN 10
+#define NUMBER_LEN 10
+#define MAX_NUMBER 999999999
+#define FILL_SIZE 233 /* so that size of one record was 1024 bytes */
+#define SKIP_OFFSET offsetof(struct data_s, checksum)
+
+static const char *Signature = "OBJ_VERIFY";
+
+POBJ_LAYOUT_BEGIN(obj_verify);
+POBJ_LAYOUT_ROOT(obj_verify, struct root_s);
+POBJ_LAYOUT_TOID(obj_verify, struct list_el);
+POBJ_LAYOUT_END(obj_verify);
+
+struct data_s {
+	char signature[SIGNATURE_LEN];
+	char number[NUMBER_LEN];
+	uint32_t fill[FILL_SIZE];
+	uint64_t checksum;
+};
+
+struct list_el {
+	struct data_s data;
+	POBJ_LIST_ENTRY(struct list_el) entries;
+};
+
+struct root_s {
+	POBJ_LIST_HEAD(verify, struct list_el) list;
+};
+
+/*
+ * list_element_constr -- constructor of a list element
+ */
+static int
+list_element_constr(PMEMobjpool *pop, void *ptr, void *arg)
+{
+	struct list_el *el = (struct list_el *)ptr;
+	long long unsigned *count = arg;
+
+	while (*count > MAX_NUMBER)
+		*count -= MAX_NUMBER;
+
+	memcpy(el->data.signature, Signature, sizeof(el->data.signature));
+	snprintf(el->data.number, NUMBER_LEN, "%09llu", *count);
+
+	for (int i = 0; i < FILL_SIZE; i++)
+		el->data.fill[i] = (uint32_t)rand();
+
+	util_checksum(&el->data, sizeof(el->data), &el->data.checksum,
+			1 /* insert */, SKIP_OFFSET);
+
+	pmemobj_persist(pop, &el->data, sizeof(el->data));
+
+	return 0;
+}
+
+/*
+ * do_create -- (internal) create a pool to be verified
+ */
+static void
+do_create(const char *path, const char *layout)
+{
+	PMEMobjpool *pop;
+
+	if ((pop = pmemobj_create(path, layout, 0, S_IWUSR | S_IRUSR)) == NULL)
+		if ((pop = pmemobj_open(path, layout)) == NULL) {
+			out("!pmemobj_open: %s", path);
+			exit(-1);
+		}
+
+	TOID(struct root_s) root = POBJ_ROOT(pop, struct root_s);
+
+	out("create(%s): allocating records in the pool ...", path);
+
+	srand((unsigned int)time(NULL));
+
+	long long unsigned count = 0;
+	while (!OID_IS_NULL(POBJ_LIST_INSERT_NEW_HEAD(pop, &D_RW(root)->list,
+							entries,
+							sizeof(struct list_el),
+							list_element_constr,
+							&count))) {
+		count++;
+	}
+
+	TOID(struct list_el) first = POBJ_LIST_FIRST(&D_RO(root)->list);
+	TOID(struct list_el) second = POBJ_LIST_NEXT(first, entries);
+
+	long size = D_RO(first)->data.signature - D_RO(second)->data.signature;
```

Comment:
You can't rely on this being the size of structure. Just use sizeof(struct list_el).